CONFETI-39  feat: [온보딩 뷰] 선택한 아티스트 조회 API 구현

jher235 opened the pull request:
## 🔊 Summaries
<!-- 본인이 한 작업을 요약해주세요. -->

- 기존 온보딩에서 이전에 조회했던 아티스트 목록을 set으로 관리하던 부분을 UserOnboardCacheDTO 객체로 변경하여 선택한(==좋아요 누른) 아티스트 목록, 조회했던 아티스트 목록을 하나의 레코드에서 관리하도록 변경
  - 변경 이유는
  1. 온보딩 진행 중에 유저 이탈
  2. 이후 해당 유저가 다시 온보딩 진입 시 좋아요를 눌렀었던 목록은 없는데, 이전에 조회했던 아티스트 목록만 redis에 남아서 제대로 온보딩이 불가할 수 있었음
  - 위 상황을 방지하기 위해 선택한 아티스트와 조회했던 아티스트 목록을 한번에 관리함 
- 선택한(==좋아요 누른) 아티스트 목록을 조회하는 API 추가

- topArtist를 가져올 때 캐시를 사용하도록 변경
  - apple-music-api:top-artists 캐시 추가
  - topArtist 를 가져올 경우 우선적으로 캐시를 확인함
  - topArtist 가 캐싱되어 있지 않아 AppleMusicApi 로 가져올 경우 캐싱하는 로직 추가
- topArtist를 가져올 때 기존에 유저가 조회했던 아티스트 목록은 제외하도록 변경

## ✨ Notification 
<!-- 참고할 사항을 작성해주세요. -->

jher235 committed:
CONFETI-39 feat: 캐시에서 관리하는 데이터 형태를 set에서 UserOnboardArtistsDTO 객체로 변경
- 캐시에서 좋아요를 누른 아티스트와 현재 조회한 아티스트의 TTL을 동일하게 관리하기 위함

jher235 committed:
CONFETI-39 feat: 온보딩 과정에서 좋아요 누른 아티스트 목록을 반환하기 위한 서비스단 DTO 작성

jher235 committed:
CONFETI-39 feat: 온보딩 과정에서 좋아요 누른 아티스트 목록을 반환하기 위한 presentation 단 DTO 작성

jher235 committed:
CONFETI-39 feat:  캐시에서 관리하는 데이터 형태를 set에서 UserOnboardArtistsDTO 객체로 변경하며 기존 로직 수정

jher235 committed:
CONFETI-39 feat: 온보딩 과정에서 좋아요 누른 아티스트 목록 조회 로직 구현

jher235 committed:
CONFETI-39 feat: Exception 필드에 @Schema(hidden = true) 추가

jher235 committed:
CONFETI-39 feat: UserOnboardControllerV4 작성
- 온보딩 진행 중 favorite 으로 선택했던 아티스트 목록 조회 api 추가

jher235 committed:
CONFETI-39 refactor: 응답 타입을 swagger 에서 자동으로 매핑하도록 수정
- docs 인터페이스를 docs 패키지 하위로 옮김

jher235 committed:
CONFETI-39 feat: @AuthErrorResponses 추가
- role 값에 의해 접근 못하는 경우가 존재하므로 추가

jher235 committed:
CONFETI-39 feat: 네이밍 변경
favoriteArtistIds -> artists

jher235 committed:
CONFETI-39 feat: cacheExposedArtist 로 메서드 네이밍 변경
- 기존 cacheTopArtist -> 변경 후 cacheExposedArtist 노출된 아티스트 id 를 추가하는 메서드이므로 위와 같이 변경함

jher235 committed:
CONFETI-39 refactor: topArtist를 가져올 때 캐시를 사용하도록 변경
- apple-music-api:top-artists 캐시 추가
- topArtist 를 가져올 경우 우선적으로 캐시를 확인함
- topArtist 가 캐싱되어 있지 않아 AppleMusicApi 로 가져올 경우 캐싱하는 로직 추가

jher235 committed:
CONFETI-39 refactor: topArtist를 가져올 때 기존에 유저가 조회했던 아티스트 목록은 제외하도록 변경
- UserOnboardService 의 getCachedExposedArtistIds 메서드에서 캐시가 존재하지 않을 경우 빈 Set을 반환하도록 방어 로직 추가
- 온보딩 캐시가 존재하지 않는 경우에도 정상적으로 사용하기 위함

ch1hyun commented on src/main/java/org/sopt/confeti/api/user/facade/UserOnboardFacade.java:
```diff
@@ -1,72 +1,103 @@
+    public UserOnboardTopArtistsDTO getTopArtists(int limit, long userId) {
+        List<ConfetiArtist> topArtists = getAllTopArtists();
+
+        Set<String> exposedArtistIds = userOnboardService.getCachedExposedArtistIds(userId);
```
ExposedArtists 라는게 클라이언트에서 선택된 아티스트를 의미하는거죠?

jher235 replied:
노출된 아티스트였어요! 
FavoriteArtists - 클라이언트가 선택한 (== 좋아요를 누른) 아티스트 
ExposedArtists - 클라이언트가 한번이라도 조회한 아티스트 
였습니다~

ch1hyun replied:
제가 이해한 대로는
1. MUSIC_TOP_ARTISTS key로 전체 Top Artist 목록 조회 (유저 상관없이 공통)
2. user id를 key로 Exposed Artist 목록, Favorite Artist 목록 조회 (유저 개인화 O)
입니다.

먼저 유저의 정상 행동 패턴으로는 다음과 같습니다.
1. 온보딩 진입
- (1)의 전체 Top Artist 목록 조회
- 클라에 아티스트 목록 갱신
2. 아티스트 선택
- (2) user id로 선택된 아티스트를 Favorite Artist에 캐싱
- 관련된 아티스트 목록 조회
- (2) 관련된 아티스트 목록을 user id로 Exposed Artist에 캐싱
- 클라에 아티스트 목록 갱신
3. 아티스트 삭제
- (3) user id로 선택된 아티스트 중 일부를 Favorite Artist 캐시에서 제거
- Exposed Artist는 무관
- 클라에서는 기존에 응답된 데이터를 가지고 UI에 반영 (API X)
4. 온보딩 종료
- Favorite Artist, Exposed Artist 캐시 제거

<br>

다음은 유저의 비정상 행동 패턴입니다.
온보딩 중 이탈 후 재진입 시나리오입니다.
1. 온보딩 진입
- (1)의 전체 Top Artist 목록 조회
- 클라에 아티스트 목록 갱신
2. 아티스트 선택
- (2) user id로 선택된 아티스트를 Favorite Artist에 캐싱
- 관련된 아티스트 목록 조회
- (2) 관련된 아티스트 목록을 user id로 Exposed Artist에 캐싱
- 클라에 아티스트 목록 갱신
3. 온보딩 이탈
4. 온보딩 재진입 (30분 경과 후, Token Expired X)
- 유저가 브라우저까지 종료해서 클라이언트에 데이터가 남아있지 않은 상황
- `GET /onboard/artists` 조회
- 이전까지 Exposed된 아티스트와 선택된 Favorite 아티스트를 응답하면 클라이언트에서 필터링 후 UI에 보여줌
5. 아티스트 삭제
- (3) user id로 선택된 아티스트 중 일부를 Favorite Artist 캐시에서 제거
- Exposed Artist는 무관
- 클라에서는 기존에 응답된 데이터를 가지고 UI에 반영 (API X)
6. 온보딩 종료
- Favorite Artist, Exposed Artist 캐시 제거

작성하신 로직의 경우 Exposed된 아티스트의 목록을 제거하고 응답하는 것으로 보여요.
이 경우 비정상 행동 패턴의 4번 과정에서 문제가 발생할 것 같아요.
그래서 로직의 흐름이 다음과 같아야 한다는 생각입니다.
1. Exposed Artist 조회
2. (1)이 있을 경우 반환, 없을 경우 (3)
3. getAllTopArtists
4. cache artists

선택된 아티스트 목록 조회 API는 따로 추가되어야할 것 같아요.

ch1hyun committed:
CONFETI-39 feat: 스웨거에 버전 업데이트 변경사항 추가

ch1hyun commented on src/main/java/org/sopt/confeti/api/user/controller/UserOnboardControllerV4.java:
```diff
@@ -0,0 +1,118 @@
+        @RequestParam(required = false, defaultValue = "100") @Min(1) @Max(200) int limit
+    ) {
+        UserOnboardTopArtistsDTO topArtists = userOnboardFacade.getTopArtists(limit, userId);
+        userOnboardFacade.cacheTopArtistsToUser(userId, topArtists);
```
내부에서 캐싱을 하고 있는 것으로 보이는데 밖에서도 하는 이유가 있으신가요?

jher235 replied:
현재 내부에서 캐싱하는 부분은 순수 TopArists 목록이여서 유저가 조회한 Artist 캐시에 TopArtist를 저장하는 부분은 존재하지 않아요. 그래서 이렇게 따로 cacheTopArtistsToUser 부분을 두었습니다.

jher235 replied:
그런데 이걸 다시 확인하면서 떠오른게,
1. 현재 cacheTopArtistsToUser 메서드를 수정 전 버전과 동일하게 사용하다보니 기존 favorite 이 지워질 수 있다는 점 
2. 온보딩 중 이탈 후 재진입 시 topArist를 이미 한번씩 조회해서 정상적인 온보딩이 힘들 수 있다는 점.. 이 있네요 
이 부분은 현재 제가 임의로 TopArtist 캐싱을 위해 limit에 맞게 가져오지 않고 Top 100 의 음악을 가져와서 TopArtist 목록을 생성 후 전부 캐싱하도록 두었긴한데, 온보딩 중 이탈이 여러번 발생하면 역시 문제가 생길 것 같아요. 그래서 온보딩 이탈 후 재진입시 보여지는 TopArtist 목록에는 favorite으로 선택한 Artist만 제외되도록 수정해두겠습니다!!

jher235 committed:
CONFETI-39 refactor: cacheOnboardArtists 메서드명 변경
- 기존 cacheTopArtists에서 cacheOnboardArtists 로 변경

jher235 committed:
CONFETI-39 feat: getTopArtists 로직 수정
- topArtist 에서 조회한 아티스트를 유저 온보딩 캐시에 추가
- TopArtist 목록에는 favorite으로 선택한 Artist만 제외되도록 수정. TopArtist 목록 조회에서 exposed artist 목록을 제외할 경우 온보딩 중 이탈이 여러번 발생한 유저는 정상적인 온보딩 진행이 어렵다고 판단하여 수정함

jher235 committed:
CONFETI-39 feat: 불필요한 디버깅 로직 삭제

jher235 committed:
Merge branch 'feat/CONFETI-39' of https://github.com/team-confeti/confeti-server into feat/CONFETI-39

jher235 commented:
4. 온보딩 재진입 (30분 경과 후, Token Expired X)
- 유저가 브라우저까지 종료해서 클라이언트에 데이터가 남아있지 않은 상황
- GET /onboard/artists 조회
- 이전까지 Exposed된 아티스트와 선택된 Favorite 아티스트를 응답하면 클라이언트에서 필터링 후 UI에 보여줌

 위 부분에서 "이전까지 Exposed된 아티스트와 선택된 Favorite 아티스트를 응답하면 클라이언트에서 필터링 후 UI에 보여줌" 이 부분이 잘 이해가 안가는데 클라에게 이전까지 Exposed된 아티스트와 선택된 Favorite 아티스트를 응답해준다는건가요??
 
 제가 생각한 로직은
 1. 클라는 이전 온보딩 진행 여부를 판단하는게 아니고, 최초 진입 시 아래 두 API 를 호출하는 구조라고 생각했습니다.
   a.  Favorite 아티스트 목록 조회
   b. TopArtist 목록 조회. (이때는 TopArtist 목록에 Favorite 아티스트는 제외함. Exposed 아티스트는 제외하지 않음. 왜냐면 최초 온보딩 진행 시 TopArtist는 전부 Exposed 되었기 때문에 온보딩 도중 이탈이 여러번 발생하면 조회할 수 있는 아티스트가 금방 부족해지므로...)
 2. 이후 기존과 동일한 플로우로 온보딩 진행. 

이렇게 생각했었는데 문제될만한 지점이 있을까요?? 비정상 행동 패턴의 4번 과정에서 발생하는 문제는 어떤 문제일까요??
추가로 현재 V4 컨트롤러에 추가한 /artists/favorite 엔드포인트가 선택된 (==favorite) 아티스트 목록 조회 API 입니다~

jher235 committed:
CONFETI-39 feat: cacheExposedArtist 로직에서 캐싱 시 favoriteArtistId 목록을 유지하도록 수정
- 기존 로직에는 exposedArtist로만 초기화해버리는 문제가 존재했음

jher235 committed:
Merge branch 'develop' of https://github.com/team-confeti/confeti-server into feat/CONFETI-39

jher235 committed:
CONFETI-39 feat: 스레드 네이밍 변경
- 추적이 용이하도록 provider와 consumer 를 구분함.
